### PR TITLE
Writeoff employees payments

### PR DESCRIPTION
**Pull request by metas-rc**
https://github.com/metasfresh/metasfresh/issues/11172

### CODE REVIEW

**Comment by metas-rc** [comment]
This is failing because I used an element in the process para  that was already added in the database I was implementing on but it is not yet in the branch. Pinged the dev who introduced it so we can sync tomorrow.

File: backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java [hunk 1]
```diff
@@ -577,6 +581,30 @@ public boolean isCashTrx(@NonNull final I_C_Payment payment)
...
+			final PaymentId paymentId = PaymentId.ofRepoId(payment.getC_Payment_ID());
+
+			final Money moneyToWriteOff = Money.of(paymentDAO.getAvailableAmount(paymentId),
+												   CurrencyId.ofRepoId(payment.getC_Currency_ID()));
+
+			logger.info("Writing off {} for the payment{}",
+						moneyToWriteOff,
+						payment);
+
+			final Date today = SystemTime.asDate();
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java:600]: take it as a parameter

File: backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java [hunk 2]
```diff
@@ -577,6 +581,30 @@ public boolean isCashTrx(@NonNull final I_C_Payment payment)
 		return tenderType.isCash();
 	}
 
+	@Override
+	public void fullyWriteOffPayments(@NonNull final Iterator<I_C_Payment> payments)
+	{
+		while (payments.hasNext())
+		{
+			final I_C_Payment payment = payments.next();
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java:589]: refactor and introduce fullyWriteOffPayment(I_C_Payment)

File: backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java [hunk 3]
```diff
@@ -577,6 +581,30 @@ public boolean isCashTrx(@NonNull final I_C_Payment payment)
...
+
+			final PaymentId paymentId = PaymentId.ofRepoId(payment.getC_Payment_ID());
+
+			final Money moneyToWriteOff = Money.of(paymentDAO.getAvailableAmount(paymentId),
+												   CurrencyId.ofRepoId(payment.getC_Currency_ID()));
+
+			logger.info("Writing off {} for the payment{}",
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java:596]: use logger.debug

File: backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java [hunk 4]
```diff
@@ -577,6 +581,30 @@ public boolean isCashTrx(@NonNull final I_C_Payment payment)
...
+						moneyToWriteOff,
+						payment);
+
+			final Date today = SystemTime.asDate();
+
+			final I_C_AllocationHdr allocationHdr = paymentWriteOff(payment, moneyToWriteOff.toBigDecimal(), today);
+
+			logger.info("C_AllocationHdr {} created for the payment {}", allocationHdr, payment);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java:604]: use logger.debug

File: backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java (hunk 1 above)
Comment (metas-rc) [reply backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java:600]: ✔️

File: backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java (hunk 2 above)
Comment (metas-rc) [reply backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java:589]: ✔️

File: backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java (hunk 3 above)
Comment (metas-rc) [reply backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java:596]: ✔️

File: backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java (hunk 4 above)
Comment (metas-rc) [reply backend/de.metas.business/src/main/java/de/metas/payment/api/impl/PaymentBL.java:604]: ✔️

File: backend/de.metas.business/src/main/java/de/metas/payment/api/IPaymentBL.java [hunk 5]
```diff
@@ -117,6 +118,8 @@
 
 	boolean isCashTrx(final I_C_Payment payment);
 
+	void fullyWriteOffPayments(Iterator<I_C_Payment> payments, Date writeOffDate);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/payment/api/IPaymentBL.java:121]: avoid Date or Timestamp in new APIs.
Consider using Instant or ZonedDateTime.

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/payment/process/C_Payment_Employee_WriteOff.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	private final IPaymentBL paymentBL = Services.get(IPaymentBL.class);
+
+	@Override
+	protected String doIt() throws Exception
+	{
+		final Iterator<I_C_Payment> paymentsForEmployees = paymentDAO.retrieveEmployeePaymentsForTimeframe
+				(p_OrgId,
+				 p_startDate,
+				 p_endDate);
+		final Date today = SystemTime.asDate();
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/payment/process/C_Payment_Employee_WriteOff.java:63]: i think it's better to introduce a new process parameter called `WriteOffDate` and set it to TODAY by default.

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments

**Comment by metas-rc** [comment]
Approved by Teo

**Pull request merged**